### ISSUE TITLE
Theme Related Problems

### ISSUE THREAD

**Issue by NABN00B** [hrydgard/ppsspp#18802]
### What should happen

Hi, I've started cooking themes yesterday and decided I should collect the problems I've encountered.

1. If an .ini file doesn't start with a definition `[` but an empty line or a comment, an empty theme will be added to the selection menu.
![image](https://github.com/hrydgard/ppsspp/assets/22328803/4f4641d8-a316-4192-bd17-5b5678857c68)

2. When a tooltip is presented after enabling an option, the text color is hardcoded to be white(-ish)
![image](https://github.com/hrydgard/ppsspp/assets/22328803/55f461ad-ff8f-4beb-a761-857a318283e9)

3. `PopupStyleBg` has no adequate Foreground option, `ItemStyleFg` is used instead.
![image](https://github.com/hrydgard/ppsspp/assets/22328803/85cbc59c-443d-4003-b1a4-08fbe641f9da)

4. `InfoStyleBg` seems to be unused (or at least I can't find it anywhere.)

5. Headers usually don't have backgrounds but the Collapsible Headers under Control remapping do and use `ItemStyleBg`. This is inconsistent behavior. There is also no customizable header background color adequate to `HeaderStyleFg`.
![image](https://github.com/hrydgard/ppsspp/assets/22328803/67e3a33b-de3a-4acc-a6e2-109b3dba88c0)

Let me know what you think.

### Who would this benefit

Theme editors

### Platform (if relevant)

None

### Games this would be useful in

-

### Checklist

- [X] Check the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already implemented.
- [X] [Search for other requests](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same feature.

**Comment by hrydgard** [comment]
1. Empty theme is a silly bug, I'll fix it right away.
2. Fixed in #18800 
3. Need to add, yes.
4. I'll clean that up. 
5. I added backgrounds to collapsible headers to emphasize that they're clickable, but it indeed needs to be themable separately.

**Comment by hrydgard** [comment]
I think we need to think a bit more about the header background, it would be weird to now just add a HeaderStyleBg but only use it for collapsible headers.

I'll fix the popup color though shortly.

**Comment by NABN00B** [comment]
3. According to the files generated by https://ppsspp-theme-editor.onrender.com/ there used to be `InfoStyleFg` at one point.

**Comment by NABN00B** [comment]
The game list settings popup seems to be another can of worms... featuring the combination of `PopupStyleBg`, and `HeaderStyleFg`. 
![image](https://github.com/hrydgard/ppsspp/assets/22328803/54339543-cc7c-4ffc-aee9-e7885f958f4a)

**Comment by hrydgard** [comment]
Yeah.. It all needs a bit of an overhaul before it's really worth trying to theme it.

**Comment by AccountGc** [comment]
Do I have to purchase games?

**Comment by hrydgard** [comment]
In #19547, I improve on this a bit (The headers in the gamelist properties get the default popup style), I changed the background of InfoStyle to look better.

**Comment by NABN00B** [comment]
> In #19547, I improve on this a bit (The headers in the gamelist properties get the default popup style), I changed the background of InfoStyle to look better.

Thanks, and sorry for the distraction!
I understand that this was meant to fix the game list settings popup only, is that correct?

**Comment by NABN00B** [comment]
Posting my opinion has been long overdue.

I played around with themes a bit in v1.14 to see the state of customizability before the [simplification](https://github.com/hrydgard/ppsspp/pull/16779) happened.  
Comparing the options and limitations of v1.14 to v1.18 helped me to formulate an idea on how I would address some of points in the original post.

(Note that whenever I mention "style" I mean a pair of FG and BG colors.)

In general, my opinion is that custom themes should be able to:
- customize the colors of as many different types of UI elements as (reasonably) possible
- *and* avoid having to recycle the colors of one type of UI element for other types.

The way I currently see it, this is achievable by:
- giving their own style to UI element types that look distinct from each other (`5`, `12`, `14`)
- *and* ensuring that each style has both a FG color and a BG color. (`3`, `4` `5`)

In particular:
- `PopupStyle` should have its `PopupStyleFg` restored (`3`)
- `PopupTitleStyle` should be restored and receive its own `PopupTitleStyleBg` (`3`)
- `InfoStyle` should preserve and have its `InfoStyleBg` actually applied (`4`)
- `HeaderStyle` should receive its own `HeaderStyleBg` (`5`)
- a new style should be introduced for collapsible headers (`5`)
- a new style should be introduced for tooltips (`12`)
- a new color (or style?) should be introduced for the scrollbar (`14`)

Currently the topbar on the Homebrew Store view is a unique exception, but as soon as we introduce another view with a topbar, a new style should be introduced for them as well.

...and if we truly want to achieve `giving their own style to UI element types that look distinct from each other`, then the various `ButtonStyles` should also be restored... (oof)

With all these changes the Default/Dark themes should still be able to keep their current look.

I'm not sure what else to write. With only @hrydgard , @iota97 and myself having worked with themes, I'm not expecting a very wide discussion, but I hope they will also post their opinions eventually.

**Comment by hrydgard** [comment]
All reasonable! Thanks for thinking it through and writing it down, it's great stuff and will make it easier for me to actually make the changes needed. I'll get all of it in before 1.19.

**Comment by iota97** [comment]
I agree that we should at least make sure every element has both an FG and a BG color. We can set them to be transparent anyway to preserve the default theme. This should be quite simple code wise.

The "every element part" is a bit more complicated. I admit I'm not following the evolution of the code base, but AFAIR many items get themed and/or use a given theme because it used to look good with the default one. That means that many things didn't use the correct semantic of `ButtonStyles` and often used a wrong one, especially if both had the same color in the default theme.

Additionally, we should have different styles based on the possibility that an element might look distinct from another given also the context. For example, a button that is inside a popup might have a different style in some other theme, however they are no different from the C++ perspective than any other button.

To summarize, the FG/BG part should be easy enough to do. The other might require a bit of design to do things right.

**Comment by hrydgard** [comment]
Maybe popups shouldn't use a different set of styles at all? Would simplify things. Although, I do think it looks good.. A bit torn on that.

**Comment by NABN00B** [comment]
> The "every element part" is a bit more complicated. I admit I'm not following the evolution of the code base, but AFAIR many items get themed and/or use a given theme because it used to look good with the default one. That means that many things didn't use the correct semantic of `ButtonStyles` and often used a wrong one, especially if both had the same color in the default theme.


I've thrown together a hideous looking theme that we could use to spot UI element types that are assigned the incorrect style.
[uistyledebugguide.ini.txt](https://github.com/user-attachments/files/17648459/uistyledebugguide.ini.txt)


> Additionally, we should have different styles based on the possibility that an element might look distinct from another given also the context. For example, a button that is inside a popup might have a different style in some other theme, however they are no different from the C++ perspective than any other button.

That is an interesting idea, but is it worth it if it requires increasing complexity? Maybe I'm overthinking it...

**Comment by hrydgard** [comment]
> Do you mean getting rid of PopupStyle completely?

Yeah, although like I said, I'm not sure it's a good idea.

**Comment by iota97** [comment]
I think having a different style helps on the visual clue that is a popup that just opened. If I selected an option and I would get a popup that blend with the background I would be quite confused. I agree it could simplify things, but welcome to UX where simple things usually are not really an option, and even small things are a lot of work to get right.

Another thing we could do is to list the different thing that need to be themed and place them all in a single screen in the developer options. It could help debugging this and can be useful later for theme creations. Ideally, we could eventually allow theme editing with live preview within PPSSPP.

**Comment by NABN00B** [comment]
About issue `17`, I believe the code [here](https://github.com/hrydgard/ppsspp/blob/489db6b92c75d9f16bc4eb3216a3bed7fe8cc34a/UI/Theme.cpp#L194) relies on `Default` theme being index `0`, which it isn't because of issue `16`.

**Comment by hrydgard** [comment]
#19984 implements some of the suggestions above. More coming.

**Comment by hrydgard** [comment]
Also added fix for `17` and a tooltip style.

**Comment by hrydgard** [comment]
Went ahead and added CollapsibleHeader style and also fixed InfoItem (though it doesn't do the half transparent when selected anymore, would need even more styles for that...).

Anyway, postponing further theme changes beyond these to 1.20 unless you want me to prioritize something.

**Comment by NABN00B** [comment]
That's a lot of changes. 🚀 
It'll take me some time to go over them to get familiar with the current state and update the existing themes.
I'll get back to you afterwards.
Thanks! 😄

**Comment by NABN00B** [comment]
> [#19984](https://github.com/hrydgard/ppsspp/pull/19984) implements some of the suggestions above. More coming.

It feels better already. 😄

I've also noticed something new -- looks like our theming of the sliders (those used in popups) is a little weird, since it relies on recycled colors. What's worse is that in #16779 we even lost the highlight on the knob when the slider is focused.

I went ahead and collected how the color usage changes between some versions.

| Element | v1.14.4 | v1.18.1 | v1.18.1-1098 |
| --- | --- | --- | --- |
| Slider | `PopupTitleStyleFg` | `ItemStyleFg` | `ItemStyleFg` |
| Slider Knob | `PopupStyleFg` | `ItemStyleFg` | `PopupStyleFg` |
| Focused Slider Knob | `PopupTitleStyleFg` | `ItemStyleFg` ❗ | `ItemStyleFg` ❗ | 

We should decide how to handle this.

**Comment by hrydgard** [comment]
Yeah, it's worth doing something about the sliders. I'll look into it tomorrow, and see if we can un-recycle something.

**Comment by NABN00B** [comment]
Another thing I found is that `PopupStyleFg` wasn't reapplied to every element it used to be applied to in 1.14.4.
In popups the text input fields and the label next to them still use `ItemStyleFg`.

Furthermore, it seems that the text input fields weren't uniformly formatted, even before 1.14.4. For example in the pro ad-hoc ip address popup `InfoStyleFg` is used.

**Comment by hrydgard** [comment]
I'm going to have to make a simple mechanism for controls to know if they're in a popup and choose styles accordingly, I think.

**Comment by hrydgard** [comment]
Yeah there are some ugly workaround around, like SetTextColor on textboxes in popups. I'm getting rid of those.

**Comment by NABN00B** [comment]
> Yeah there are some ugly workaround around, like SetTextColor on textboxes in popups. I'm getting rid of those.

Looks great after #19997 -- marked them done in the original post.
Thanks!

**Comment by NABN00B** [comment]
> Yeah, it's worth doing something about the sliders. I'll look into it tomorrow, and see if we can un-recycle something.

I went ahead and took a look myself. The colors are set in [`Slider::Draw`](https://github.com/hrydgard/ppsspp/blob/d36230937d0b72986559b971a36057f15158eae3/Common/UI/View.cpp#L1515) and [`SliderFloat::Draw`](https://github.com/hrydgard/ppsspp/blob/d36230937d0b72986559b971a36057f15158eae3/Common/UI/View.cpp#L1644C6-L1644C23). Sliders are only ever used in popups, which simplifies things a lot.

The line of the slider seems easy enough to customize:
- Introduce `sliderStyle`, where `sliderStyleFg` is the filled part and `sliderStyleBg` is the unfilled part of the bar. (We might as well make the unfilled part customizable too right now while we're looking at sliders.)

The knob could be dealt with in a couple of ways:
- Introduce a single `sliderKnobStyle`, where `sliderKnobStyleFg` is the focused color and `sliderKnobStyleBg` is the unfocused color of the knob. A bit hacky and unintuitive.
- Introduce separate `sliderKnobStyle` and `sliderKnobFocusedStyle`, where `*Fg` is the respective knob color for each scenario, and `*Bg` might or might not be used for accessibility purposes later? I'm not sure if this makes any sense...
- Introduce standalone `sliderKnobColor` and `sliderKnobFocusedColor` values. Plain and simple.

**Comment by hrydgard** [comment]
I can throw something in that just lets you set the slider color. Then we can set alpha to 1.0 when dragging, or have a separate color for that. We just need to decide :)

**Comment by NABN00B** [comment]
> Then we can set alpha to 1.0 when dragging, or have a separate color for that. We just need to decide :)

Off the top of my head I can't imagine how that that would look like.

I dislike the idea of using hardcoded alpha values in general.

Even where alpha manipulation is already used, it would allow for more customization if we treated alpha relative to the theme colors' defined alpha rather than a hardcoded percentage.

**Comment by iota97** [comment]
Im a bit confused, if there is a color for the items in general and one when they are focused already shouldn't just we use those to for the slider knob? Unless there is a reason to change only the knob in the theme ofc :)

**Comment by NABN00B** [comment]
> Im a bit confused, if there is a color for the items in general and one when they are focused already shouldn't just we use those to for the slider knob? Unless there is a reason to change only the knob in the theme ofc :)

Oh, now I understand it. What you said makes a lot of sense actually.

So in summary:
- use `ItemFocusedStyleBg` for knob, when knob is selected
- use `` for knob, while slider is being moved / knob is held down; and reapply `ItemFocusedStyleBg` after this is no longer true
- need a new `sliderKnobColor` value for idle knob
- need 1 (or 2?) colors for filled slider bar (and unfilled slider bar?)

Is that correct? 🤔

**Comment by iota97** [comment]
I would try to reduce the number of elements. In my mind:

- the default knob should just have the same color of a button as it is also a selectable action item.
- focused and selected should also match buttons by the same logic.
- The filled bar can have the same color of the knob (even better if it changes when focused and selected as the whole slider is the object not only the knob).
- The unfilled part is kinda of the disabled part so I would use the color disabled items uses.

I don't have a strong opinion and adding new things is fine too.

**Comment by NABN00B** [comment]
>     * the default knob should just have the same color of a button as it is also a selectable action item.
>     * The unfilled part is kinda of the disabled part so I would use the color disabled items uses.

Usually these values are semi-transparent though. It might look weird for the sliders, especially the filled part as that would be laid on top of 2 (usually) other semi-transparent elements (popup, unfilled part).

**Comment by NABN00B** [comment]
If we consider the slider to be the same kind of GUI item as for example buttons or settings list item (idk what these are called), then we need to look at how sliders look like when they're not in a pop-up.

**Comment by hrydgard** [comment]
It's a good point - though sliders are currently always in popups, it might not stay that way forever.

Anyway, sorry for the confusion earlier, I was thinking about scroll handles when I made my alpha comment :) The sliders are trickier, yeah.

**Comment by NABN00B** [comment]
> Anyway, sorry for the confusion earlier, I was thinking about scroll handles when I made my alpha comment :) 

Oh, I see now. Though it being held down is already being indicated by increasing its thickness.

**Comment by NABN00B** [comment]
I've tried to come up with a potential solution and the results are... interesting.

Changes:
https://github.com/NABN00B/ppsspp/pull/11/files
![Image](https://github.com/user-attachments/assets/f62b1b9a-34ef-4221-b5c8-8c692ade36b1)

Explanation:
- I've combined slider bar and knob colors into one. Provides better feedback if both are the same color. (The user's finger blocks the view of the knob on touch screens.)
- If a slider were to be used outside popups, it would be surrounded by `ItemStyle` / `ItemFocusedStyle` / `ItemDownStyle` background. For this reason the slider needs to use the foreground colors when focused / held down outside popups.
- Popup background colors can vastly differ from that of `ItemStyle`, so using the same foreground colors would be a huge limitation on popup background. For this reason I made the slider use the background colors when focused / held down inside popups.

Win x64 Build:
https://github.com/NABN00B/ppsspp/actions/runs/14644802685/artifacts/3002987052

This implementation (apart from looking ugly with the 2 default themes popup bg colors) still ends up having the slider colors as a limiting factor in some of my themes. :\

A potential improvement of this could be having dedicated colors for idle and focused sliders **inside popups**.
Outside popups the implementation should work the same. (I think?)

Alternatively we need to keep looking for other solutions.

P.S.: Since System Information -> Internals contains various visual test items now, can @hrydgard try placing a slider there inside an item so we can see how it looks?

**Comment by NABN00B** [comment]
> Anyway, sorry for the confusion earlier, I was thinking about scroll handles when I made my alpha comment :)

Here's what it looks like:
![Image](https://github.com/user-attachments/assets/1024ec6d-992a-4a9f-9632-cb9b256f3829)

**Comment by NABN00B** [comment]
> Hm, that doesn't look very good when fully opaque :)

I think it's just the white color. I do have a theme that uses fully opaque slider color and it looks pretty decent.

**Comment by hrydgard** [comment]
Well, my main problem with it is how it interacts with the checkboxes (touching them). We'd need to shrink the bar slightly or move them left a tiny bit, then it'd be ok.

**Comment by NABN00B** [comment]
> Well, my main problem with it is how it interacts with the checkboxes (touching them). We'd need to shrink the bar slightly or move them left a tiny bit, then it'd be ok.

The checkboxes aren't even the biggest problem. It is the right end of strings that are being covered up by it. But yeah, it needs some adjustments one way or another...

**Comment by NABN00B** [comment]
I've updated the original post so that the screenshots are hidden, making the list actually readable. 🧠 

Let's summarize the remaining points in order of severity:
- `15` and `20` are bugs
- `11` could be a bug
- `9`, `10`, `12` and `21` require messing with transparency stuff
  - `21` requires adjusting either the scrollbar or the right-alignment of the contents of menu items
- `16` is not an issue, but requires hrydgard to make a decision on whether it's worth defining the Default theme in 2 places
- `4` requires thinking on whether and how we want to use the background color

Overall themes are in a really good state right now, a lot has been fixed and improved in the last 2 major releases. 👍

**Comment by iota97** [comment]
For `16` I defined it twice as they have different purpose:

- the hard-coded value in the C++ file are used as fallback if a style is not provided for one entry (you only need to provide the styles entry that are different from the default). This is useful to mitigate breakage of old theme if a new entry is added as the default one will be used at least. It might be possible to read those value from the ini, but that was gonna be bit complicated code wise iirc. Also, hard-coding help providing a style if the ini is not found for some weird filesystem access reason (without this even showing an error might end up black on black and useless).

- the ini version was provided for reference, back then we only had the default theme, having an example helps user to create their own. To avoid duplicates it override the hard-coded default one when found the ini. If you redefine it, I thinks its a consistent behaviour that it change the default theme in the app.

While the situation changed a bit since then I am not sure its worth the effort of changing it unless a "real" issues appears :)

**Comment by NABN00B** [comment]
@hrydgard I've noticed that 2 elements are themed incorrectly since the big UI overhaul:
- `22.` the version label on the main screen below the PPSSPP typeface now uses `ItemStyle` (was previously `InfoStyle`)
<img width="340" height="120" alt="Image" src="https://github.com/user-attachments/assets/f23bd396-1404-4306-bd87-ca266108f359" />


- `23.` the title of the currently selected tab in the Settings screen and Developer Tools is themed separately from other screen titles -- I think this type of element should maybe receive its own style (fg + bg), maybe `TitleStyle`?
<img width="527" height="166" alt="Image" src="https://github.com/user-attachments/assets/8dbfd98c-8e9c-4aa8-b3b2-4a5f6630e3f1" />

<img width="363" height="85" alt="Image" src="https://github.com/user-attachments/assets/9d6d2aab-781a-496a-a5bd-3f08a73bde6b" />

**Comment by NABN00B** [comment]
- `24.` The descriptions for options use `InfoStyle`, which, except for System information, is never used with a background. The usage of `InfoStyle` here leads to unexpected color combinations. The background uses `ItemStyleBg`.

I think in this context, the description should be considered as an extension of the UI item and recycle `ItemStyle` for consistency.

<img width="726" height="459" alt="Image" src="https://github.com/user-attachments/assets/25bc6452-8f18-4101-982b-e0bfc3f93e52" />

**Comment by hrydgard** [comment]
Yeah, that's a miss. I'll fix it
